tools/DevDiagnostics/DevHome.DevDiagnostics/Helpers/LoaderSnapAssistantTool.cs

```diff
@@ -0,0 +1,193 @@
+                        _loaderSnapFailures.Remove(info.pid);
+
+                        _dispatcher.TryEnqueue(() =>
+                        {
+                            var insight = new SimpleTextInsight();
+                            insight.Title = string.Format(CultureInfo.CurrentCulture, CommonHelper.GetLocalizedString("InsightProcessExitMissingDependenciesIdentifiedTitle"), info.processName, info.pid);
+                            insight.Description = loadersnapError;
+                            _insightsService.AddInsight(insight);
+                        });
+                    }
+                }
+            }
+        };
+    }
+
+    private void MyETWListener()
+    {
+        using TraceEventSession session = new TraceEventSession("LoaderSnapAssistantSession");
+        session.StopOnDispose = true;
+
+        // Enable the loader snaps provider
+        session.EnableProvider(WindowsImageETWProvider, TraceEventLevel.Error, LoaderSnapsFlag);
+
+        // We don't care about a lot of the ETW data that is coming in, so we just hook up the All event and ignore it
+        session.Source.Dynamic.All += data => { };
+
+        // None of the loadersnap events are handled by the TraceEventParser, so we need to handle them ourselves
+        session.Source.UnhandledEvents += UnHandledEventsHandler;
+        session.Source.Process();
+    }
+
+    private void UnHandledEventsHandler(TraceEvent traceEvent)
+    {
+        if (traceEvent.EventName.Contains("Opcode(215)"))
+        {
+            byte[] loaderSnapData = traceEvent.EventData();
+            string s = System.Text.Encoding.Unicode.GetString(loaderSnapData.Skip(10).ToArray());
+            s = s.Replace("\n\0", string.Empty);
+            s = s.Replace("\0", ": ");
+            if (s.Contains("LdrpProcessWork - ERROR: Unable to load"))
+            {
+                lock (_loaderSnapFailures)
+                {
+                    // At this point, we don't have the faulting process's executable name. Wait until we get the callback
+                    // from our service that tells of the process termination, and then we'll raise the insight.
+                    _loaderSnapFailures.Add(traceEvent.ProcessID, s);
+                }
+            }
+        }
+    }
+
+    private bool IsLoaderSnapLoggingEnabledForImage(string imageFileName)
+    {
+        // Check if the following registry key exists
+        // Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows NT\CurrentVersion\Image File Execution Options\imageFileName
+        // TracingFlags = 0x4
+        RegistryKey? key = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\Microsoft\Windows NT\CurrentVersion\Image File Execution Options\" + imageFileName, false);
+
+        if (key is null)
+        {
+            return false;
+        }
+
+        if (key.GetValue("TracingFlags") is not int tracingFlags)
+        {
+            return false;
+        }
+
+        if ((tracingFlags & 0x4) != 0x4)
```

Comment:
Great opportunity for a 

[Flags]
enum TracingFlags : uint
{
    Foo = 0x4,
    etc = 0x8,
}

Followed by
tracingFlags.HasFlag(Foo) instead of this manual stuff :)

Reply:
Example on how to do this (save you loading a web page if you don't know how)

https://github.com/microsoft/devhome/blob/user/scodar/imageoptions/tools/DevDiagnostics/DevHome.DevDiagnostics/Helpers/ImageFileExecutionOptions.cs

Reply:
Got it. I ended up defining the flags twice, as I didn't want to reference a common assembly in my elevated process. But this seems to work well.

Reply:
What's wrong with using common assemblies in an elevated process?

Reply:
Just need to worry about statics.... make sure code isn't executing unexpectedly.

Reply:
I don't see how that's ever going to work at scale.  Adding a bad project reference will get overlooked 99 times out of 100 without automation.